hazelcast/src/main/java/com/hazelcast/jet/impl/JobResult.java

```diff
@@ -37,28 +40,34 @@
 import static com.hazelcast.jet.impl.util.Util.toLocalDateTime;
 import static java.util.concurrent.CompletableFuture.completedFuture;
 
-public class JobResult implements IdentifiedDataSerializable {
+public class JobResult implements IdentifiedDataSerializable, Versioned {
```

Comment:
Just to double check: This class can't be transferred to/from the client, right?

Reply:
It is not used in client-accessible APIs. It is only stored in one of Jet IMaps. However, Management Center seems to use it via internal Jet classes in `JobManager` (it reads internal Jet maps). Is that a problem?

Reply:
If the writer or reader of this class is open-source, then the version checks here will not work as intended. The version is only written to the identified header in the enterprise edition, and on the reading part, we simply skip them in open source.

https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java#L155

https://github.com/hazelcast/hazelcast-enterprise/blob/master/hazelcast-enterprise/src/main/java/com/hazelcast/internal/serialization/impl/EnterpriseDataSerializableSerializer.java#L276-L283

When the class is only used between members, version checks are fine, as only the same minor versioned members can connect or we might be doing rolling upgrade and then all the servers would be enterprise anyway.

But, when the clients are involved in the picture, since there is no version/open-source/enterprise constraints, the version checks might not work

Reply:
will this be a problem: https://github.com/hazelcast/management-center/blob/master/src/main/java/com/hazelcast/webmonitor/service/jet/JobManager.java#L105?

@viliam-durina WDYT?

Reply:
Assuming MC uses hazelcast-enterprise under the hood, they will have no problem reading the version when it is set. 

However, if the cluster that they are reading the JobResult from is open-source, then the version will be set to UNKNOWN https://github.com/hazelcast/hazelcast-enterprise/blob/master/hazelcast-enterprise/src/main/java/com/hazelcast/internal/serialization/impl/EnterpriseDataSerializableSerializer.java#L153-L158, and according to doc, `UNKNOWN.isGreaterOrEqual` always returns `false`, so they won't be reading the `userCancelled` field. If the cluster is enterprise, that will be fine, MC will read or skip reading the field according to the version correctly.

I am assuming MC won't be writing any `JobResult` to the cluster. If they are writing though, it can be problematic. I assume older MC versions are compatible with newer Hazelcast versions, then a `JobResult` written by 5.2 MC will not include this new field. And, if we are using an open-source cluster, we will set the version to the current cluster version https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java#L117 https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/DataSerializableSerializer.java#L59 and if we are 5.3 or newer, we will always try to read this field since we will pass the version check. Then, the deserialization will fail

Reply:
MC only deserialises `JobResult`, it does not update it.

MC depends on `hazecast-enterprise` JAR. So unless it does something tricky with serialization in case of usage without license (I think it can be used with some limitations) it seem it should be able to deserialize `JobResult`. However MC 5.3 will not be able to show `userCancelled` flag for OS 5.3 cluster.

We will have to keep conditional deserializalition for entire 5.z line (if there will be more versions) if MC 5.z is to be compatible with all 5.z cluster versions.

Reply:
It is fine then. I don't think they are doing anything special, since you don't have to have a license key to use enterprise client.

Yes, they won't be able to get the accurate value if the cluster is OS, that is a bit sad, but nothing we can do about it.

It seems all MC versions are compatible with all cluster members. https://docs.hazelcast.com/management-center/5.2/getting-started/overview#hazelcast-cluster-compatibility So, right we might need to keep this check

Reply:
actually there is a problem: MC will try to deserialise `JobResult` from OS cluster using UNKNOWN version. So if cluster is 5.3 it will skip the new field `userCancelled` and one stray byte will remain in buffer (this will be true also for MC 5.3). This might work if nobody reads more from this buffer (this may be the case for `IMap.get`, I did not test it) but is very fragile.

Reply:
I don't know about MC but clients are forward compatible as well, so at least in theory it can work with newer versions too.

The problem you mentioned would likely to happen if the JobResult was part of another identified object but since it is not the case, we should be fine (for now :) ). 

We have the connected server version information (we don't use it for now, but it is available as part of the authentication response https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/ClientAuthenticationCodec.java#L173). Maybe we can make use of it

Reply:
yes, clients are forward compatible, but MC does not use client API to access `JobRecord`. It reads it directly from IMap so usual client-compatibility mechanisms are not used.

`JobRecord` is not exchanged with clients in any requests or responses.

MC has `HazelcastInstance` object, from which it can obtain cluster version. So MC could do something like this:
```java
    if (client.getHzClient().getCluster().getClusterVersion().isGreaterOrEqual(Versions.V5_3)) {
        // use actual client API (eg. new method in Job with new request underneath)
        // instead of hacking into internal IMaps
        builder.completionTime(job.getCompletionTime());
    } else {
        // use the old way for older clusters
        JobRepository jobRepository = new JobRepository(client.getHzClient());
        JobResult jobResult = jobRepository.getJobResult(job.getId());
        if (jobResult != null) {
            builder.completionTime(jobResult.getCompletionTime());
        }
    }
```

Reply:
I didn't get it. `client.getHzClient()` is `HazelcastInstance` but it is `HazelcastClientInstanceImpl`. So, when it reads from the IMap named `__jet.results`, it uses client APIs and if I am not missing anything, `JobResult` is being transferred to the client. 

Also, that method of getting the cluster version is not implemented for client https://github.com/hazelcast/hazelcast/blob/74134e54b476ce07f4cf47c8682b43d9090e573f/hazelcast/src/main/java/com/hazelcast/client/impl/proxy/ClientClusterProxy.java#L91

Reply:
`IMap.get` returns a blob which is deserialized to `IdentifiedDataSerializable` object on client side (in this case MC). But the problem is that MC configures `HazelcastClientInstanceImpl` with `versionedSerializationEnabled=false`. This causes that deserialization of IDS on MC side uses `DataSerializableSerializer` instead of `EnterpriseDataSerializableSerializer`.

> MC depends on hazecast-enterprise JAR. So unless it does something tricky with serialization

I was wrong. MC actually does something tricky - disables version support. `DataSerializableSerializer` ignores version in serialized data. Instead it uses MC version and not the connected cluster version.

All this means that if we add field in `JobResult`, MC 5.3 connected to cluster 5.2 (OS or EE - does not matter) will assume that the data is in 5.3 format and will expect additional field and fail with `EOFException` during deserialization.

Reply:
This also means that every class that is deserialized by MC directly, not by using client protocol, cannot use `Versioned` and can never be changed once MC started using it.

Reply:
BTW, this affect also `JobConfig` which is stored in `JobResult`

Reply:
I could change added code in `read` method in `JobRecord` as follows:
```java
        if (in.getVersion().isGreaterOrEqual(Versions.V5_3)) {
            try {
                userCancelled = in.readBoolean();
            } catch (EOFException e) {
                // ignore
                // this probably means that MC 5.3 tries to read JobRecord from cluster < 5.3
            }
        }
```
I don't like it, but it should work in MC usage (with `IMap.get()`). When invoked in the cluster we should never get EOF because version will be sensible.

Reply:
I see, yes not the nicest solution but I also think that it will work.

Reply:
We don't have too many options here. Either make MC not use this class or do something like this

Reply:
https://hazelcast.slack.com/archives/C9S7BV74Z/p1674565097965589?thread_ts=1674495262.742189&cid=C9S7BV74Z If this is the case, then we might not need it

Reply:
I implemented workaround. I understand that we try not to break compatibility of MC with older clusters if possible. The workaround is also needed for EE without rolling upgrade license.
For decoupling MC from JobRecord I created a separate issue: HZ-1990